OpenRA.SourceGenerators/Sync/SyncHashCodeSourceGenerator.cs

```
@@ -0,0 +1,106 @@
+	{
+		public void Initialize(IncrementalGeneratorInitializationContext context)
+		{
+			var classDeclarationsProvider = context.SyntaxProvider
+				.CreateSyntaxProvider(
+					predicate: static (syntaxNode, _) => syntaxNode.IsPartialClass(),
+					transform: static (generatorContext, _) => GetClassDeclarationSymbol(generatorContext))
+				.Where(x => x != null)
+				.Collect();
+
+			var compilationAndClasses = context.CompilationProvider.Combine(classDeclarationsProvider);
+
+			context.RegisterSourceOutput(compilationAndClasses,
+				static (context, source) => GenerateCode(source.Left, source.Right, context));
+		}
+
+		static ClassDeclarationSyntax GetClassDeclarationSymbol(GeneratorSyntaxContext context)
+		{
+			if (context.SemanticModel.GetDeclaredSymbol(context.Node) is INamedTypeSymbol classSymbol && classSymbol.ImplementsISync())
+				return context.Node as ClassDeclarationSyntax;
+
+			return null;
+		}
+
+		static void GenerateCode(Compilation compilation, ImmutableArray<ClassDeclarationSyntax> classes, SourceProductionContext context)
+		{
+			if (classes.IsDefaultOrEmpty)
+				return;
+
+			foreach (var classDeclaration in classes.Distinct())
+			{
+				var (filename, content) = ProcessType(classDeclaration, compilation);
+				if (content != null)
+					context.AddSource(filename, content);
+			}
+		}
+
+		static (string FileName, string Content) ProcessType(ClassDeclarationSyntax classDeclaration, Compilation compilation)
+		{
+			if (classDeclaration.Parent is not NamespaceDeclarationSyntax namespaceDeclaration)
+				return (null, null);
+
+			var semanticModel = compilation.GetSemanticModel(classDeclaration.SyntaxTree);
+			var typeSymbol = semanticModel.GetDeclaredSymbol(classDeclaration) as ITypeSymbol;
+			var syncedClassMembers = typeSymbol.GetAllMembers(x => x.HasSyncAttribute()).ToArray();
+			var hashCodeStrings = new List<string>();
+			var syncsTargets = false;
+			foreach (var member in syncedClassMembers)
```

Dev experience thought: If you haven't tagged any members then maybe we should throw an error to make the dev aware. This could be wrapped in a helper, e.g.
```c#
int ISync.GetSyncHash()
{
    throw Sync.NoSyncMembers();
}
```
```c#
NotSupportedException NoSyncMembers()
{
    return new NotSupportedException("descriptive error about using sync tag here");
}
```

---

I believe the lint rules should already handle that?
Turning it into a proper analyzer sounds fun though.